Separate Line reader into a reader and a scanner

kvch opened the pull request:
### Architecture
The existing `Line` reader is separated into two smaller units: a reader and a scanner. The reader decodes and converts the input bytes into UTF-8. Then the scanner looks for new lines in the output of the decoder. If it finds one, returns the new line along with its size.

### Challenges
1. Size of the encoded line to store the correct state
    As the scanner is encoding-agnostic, the length of the encoded bytes is unknown when scanning for new lines. However, it's required otherwise the state becomes corrupt. To eliminate this problem I added a helper slice to track bytes sizes. See more in the section "Size calculation".
2. Error handling
    Error detection during converting bytes is hard. Our previous solution was incorrect and hadn't caught every error. No error is reported if a users misconfigures the `encoding` of an `input`. We could use external libraries which guess the correct encoding, but it would not add as much value as it complicates the code and degrades performance. Also, we haven't heard about many problems related to selecting the correct encoding for an input file, so I think we can live without it.
3. Performance
    See more in the dedicated section.

### Size calculation
To calculate the original size of the encoded string I introduced an array. The length of the array equals the number of bytes of the converted string. The i-th element of the array shows how many bytes long in UTF-8 is the original i-th byte. In order to calculate the length of an n bytes long converted string, sum the numbers in the array from the left to right until it equals the `len` of the converted string. The original size is the number of elements that have to be added.
#### Example
String: `"tél"` (It means winter in Hungarian. Congrats, you've just learnt something new!)
Len in UTF-8: 5
```
len("tél") # equals 5
```
Latin1
```
[1 2 1 1]
count: 4
```
Original length of the encoded string: 4 bytes

UTF-8
```
[1 0 2 1 1]
count: 5
```
Original length of the encoded string: 5 bytes

UTF-16
```
[0 1 0 2 0 1 0 1]
count: 8
```
Original length of the encoded string: 8 bytes

### Possible optimizations
This is a general converter which transforms everything byte-by-byte. This could add a significant performance penalty. To avoid bad performance, specialized decoders could be introduced in later PRs.
1. No conversion and skip size calculation when encoding is UTF-8.
2. Convert the whole byte slice and skip size calculation when encoding is ASCII.
3. Convert two bytes at a time when encoding is UTF-16

kvch commented:
Still fighting with the CI.

urso commented on filebeat/reader/line/line.go:
```diff
@@ -18,177 +18,258 @@
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []int
```
let's safe some space and use `[]uint8`. On 64bit architectures an int takes 8bytes.

urso commented on filebeat/reader/line/line.go:
```diff
@@ -18,177 +18,258 @@
+		start := r.encodedBuf.Len()
+		n, err := r.in.Read(b[start:])
+		if err != nil {
+			return len(buf[:n]), err
```
this is equivalent to `return n, err`. But why return `n` here? No contents has been copied to buf yet.

urso replied:
This return does not look correct. Its equivalent to `return n, err`. We will need a global variable tracking the actual number of bytes copied into buf.

urso commented on filebeat/reader/line/line.go:
```diff
@@ -18,177 +18,258 @@
-		n, err := r.reader.Read(buf)
+		if start > 0 {
+			enc, _ := r.encodedBuf.Collect(start)
+			b = append(enc, b[start:]...)
```
this is somewhat unfortunate, as we will always allocate and copy if encodedBuf is not empty.

urso commented on filebeat/reader/line/line.go:
```diff
@@ -18,177 +18,258 @@
+				r.offset += nBytes
+				r.bytesOffset += nProcessed
+				return nBytes, nil
+			}
```
can we hist ErrShortDst here as well?

urso commented on filebeat/reader/line/line.go:
```diff
@@ -18,177 +18,258 @@
+	offset         int
+	bytesOffset    int
+	lastMessageLen int
+}
```
types should be at the top of the source file. Given both readers are somewhat non-trivial consider splitting them into separate files.

kvch replied:
yes, i kind of expected this request, so i put the types like this, so it would be easier to copy them into different files.

kvch commented:
The failing tests in Jenkins are unrelated. In Travis everything is green.

kvch commented:
I am keeping the label in progress on this PR until FF. I would rather not release this refactoring without the additional optimizations.

urso commented:
The change is not that trivial. Can you have filebeat running a few hours with this PR and monitor memory usage? Do not look at RSS, but check for `Alloc`, `HeapInuse` as provided by the go runtime (see runtime.Memstats). When plotting these, the graphs will be quite jittery. Apply some smoothing filter (e.g. moving average) and check that we did not introduce a leak (it can take a while to discover a leak, due to gc related noise).

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+type decoderReader struct {
```
some docstring about decoderReader transforming the input stream into UTF-8, by using a configurable character encoding would be nice here.

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+	b := make([]byte, r.bufferSize)
+
+	if r.buf.Len() != 0 {
+		return r.copyToOut(buf)
```
nit: what if `buf.Len()` == 1? In this case we would copy only one byte and return. It's not incorrect, but a little inefficient. Let's add a `TODO` comment at least, noting that we should try to collect at least `len(buf) - r.buf.Len()` bytes, if `len(buf) > r.buf.Len()`.

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+		n, err := r.in.Read(b[start:])
+		if err != nil {
+			return len(buf[:n]), err
+		}
```
this still looks like a potential bug. 

Let's assume before entering the loop we have this initial state:

```
encodedBuf.Len = 0
file-read-offset: 0
r.buffer_size: 1000 // len(b) == 1000
b[0:1000] = "\0..."
buf[0:1000] = "\0..."
```

The file state will contain 26 charatecters:
```
"abcdefghijklmnopqrstuvwxyz"
```

Given this state we already 

Now the read fails, but some bytes have been copied into the "internal" buffer b. Return values:
```
n = 10
err = "ups"
```
The new state is:

```
encodedBuf.Len = 0
file-read-offset: 10
b[0:10] = "abcdefghij"
b[10:1000] = "\0..."
buf[0:1000] = "\0..."
```

Next calling 'return' with this state we tell the caller `n` bytes have been written to buf -> return values are:

```
n = 10
err = "ups"
buf[0:10] = "\0..." // <- bug 1, we return contents, without having copied anythin
```

That is no content has been copied at all, yet we tell the next reader in the chain to process 10 characters.

Worse, if the next reader tries to read more, we have the initial state for the current function call at:

```
encodedBuf.Len = 0
file-read-offset: 10
r.buffer_size: 1000 // len(b) == 1000
b[0:1000] = "\0..."
buf[0:1000] = "\0..."
```

This time the read succeeds. As the filter read pointer is already at offset 16, we will read 16 bytes only: ->

```
n = 16
err = 0
b[0:16] = "klmnopqrstuvwxyz"
```

Assuming we can convert by just copying (all characters in lower ASCII range) the function return will be:

```
n = 16
err = nil
buf[0:16] = "klmnopqrstuvwxyz"
```

Due to the error handling we made the callee process the result byte sequence:
`"\0\0\0\0\0\0\0\0\0\0klmnopqrstuvwxyz"`, but the expected byte sequence should have been `"abcdefghijklmnopqrstuvwxyz"`.


Another potential bug arises we run the loop twice, with the first read succeeding and the second failing. Assuming `n=5` in the first run and `n=2` in the second run (plus error), then we will return `n=2, err="ups", buf[0:2]="ab"`. The other 5 bytes that have been writte from the file are ignored.

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+			}
+			return 0, err
+		}
+		return nBytes, nil
```
The for loop always returns. In this case the for loop can be removed, no?

The idea of the for-loop was to continue reading if there is still some more space in `buf`.

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+// msgSize returns the size of the encoded message on the disk
+func (r *decoderReader) msgSize(symlen []uint8, size int) (int, []uint8, error) {
```
what does `size` mean? Is `size` the input in number of `rune`s or number of `byte`s? Given the loop itself, I expect `size` to be the number of bytes and the return value being the number of `rune`s.

What if `size` becomes negative?

kvch replied:
It is the number of converted bytes. I renamed the variables and added more comments.

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+	var err error
+	nProcessed := 0
+	decodedChar := make([]byte, 64)
+	r.symlen = make([]uint8, len(in))
```
is there a chance `len(r.symlen) > 0` when entering `conv`?

kvch replied:
No, it's always retrieved by the scanner.

urso commented on filebeat/reader/line/decoder.go:
```diff
@@ -0,0 +1,166 @@
+		j := i + 1
+
+		for j <= srcLen {
+			nDst, nSrc, err := r.decoder.Transform(decodedChar, in[i:j], false)
```
ouch... checking `x/text`, there doesn't seem to be a `decodeRune` for any encoding. All encodings have the decode and copy functionality built right into the loop implemented by Transform :(
 
E.g. with `decodeRune` we could write the copy loop like this:

```
...
for i := 0; i < len(in); i++ {
  r, width, err := encoding.DecodeRune(in)
  if err != nil {
    ... // e.g. ErrShortSrc
  }

  uft8Width := utf8.RuneLen(r)
  if len(to) < utf8Width {
    ... // output buffer full, return
  }

  r.symLen[nProcessed] = width
  nProcessed++
  utf8.EncodeRune(to[nProcessed:], r)
}
...
```

We should extract the inner loop to decodeEncRune:

```
func decEncRune(decoder ..., in []byte) (rune, int, err) {
  for j := 0 ; j < len(in); j++ {
}
```

kvch replied:
Yes, this is the part of the code I am most worried about. It's complex and does not perform well. I would like to drop this algorithm in favour of the newer idea we discussed previously.

kvch commented:
I rebased the PR and addressed the review notes. Now I am focusing on the alternative "segmenting" algorithm we discussed. After that's complete I would like to compare the two solutions regarding complexity and performance.

urso commented on filebeat/reader/readfile/decoder.go:
```diff
@@ -0,0 +1,169 @@
+	b := make([]byte, r.bufferSize)
+	start := r.encodedBuf.Len()
+	n, err := r.in.Read(b[start:])
+	if err != nil {
```
If `n > 0`, then we must process all bytes from `b[start:start+n]`. Independent of the actual value of err.

urso commented on filebeat/reader/readfile/decoder.go:
```diff
@@ -0,0 +1,169 @@
+			return 0, err
+		}
+		r.encodedBuf.Reset()
+		b = append(enc, b[start:]...)
```
more correct/efficient would be `b = append(enc, b[start:start+n]...)`. If `n < len(b[start:])`, then this append operation will add loads of '\0' to the end of b.

urso commented on filebeat/reader/readfile/decoder.go:
```diff
@@ -0,0 +1,169 @@
+	for i < srcLen {
+		j := i + 1
+
+		for j <= srcLen {
```
if we use this approach, then this inner loop should be refactored into a `decodeRune` function/method (somewhere burried in the comments). Reading and understand this loop takes time. Code should be more skimmable at best.

urso commented on filebeat/reader/readfile/scanner.go:
```diff
@@ -0,0 +1,80 @@
+	for !separatorFound(idx) {
+		b := make([]byte, s.bufferSize)
+		n, err := s.in.read(b)
+		s.buf.Append(b[:n])
```
another place where support for pre-allocating and accessing the available write-buffer in streambuf would be nice. Like:

```
s.buf.Ensure(s.bufferSize)
n, err := s.in.inread(s.buf.WriteBuffer())
```

Btw. we shoud not use streambuf here. The Reset method does set the read/write pointer to 0, but the underlying slice is not reset (buffer is reused). On big line/allocation can grow the buffer and waste memory we are never able to return.

kvch commented:
Closing as the solution in https://github.com/elastic/beats/pull/7940 performs better.
